### PR TITLE
Assignment-7+8-PairWork-Hadi-Rahat

### PR DESCRIPTION
**Pull request by HadiM2**
@rahat15
@HadiM2

### CODE REVIEW

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        // 1.
+        public IEnumerable<string> CsvRows
+        {
+            get
+            {
+                using (StreamReader streamReader = new StreamReader(FILEPATH))
+                {
+                    streamReader.ReadLine();
+
+                    while (streamReader.EndOfStream)
```
Comment (Aramirez61) [inline Assignment/SampleData.cs:20]: I think this should be changed to while (!streamReader.EndOfStream)
because the file should keep getting read until it is NOT the end of stream, then exit once it reaches it

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                    {
+                        yield return streamReader.ReadLine()!;
```
Comment (Aramirez61) [inline Assignment/SampleData.cs:22]: I believe you need to use the linq method .skip(1) after readline to skip the header of the csv file

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Assignment
```
Comment (Aramirez61) [inline Assignment/SampleData.cs:6]: Use file scoped namespace

File: Assignment/Assignment.csproj [hunk 4]
```diff
@@ -1,5 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
-
+    <PropertyGroup>
+	  <TargetFramework>net8.0</TargetFramework>
+	  <LangVersion>12.0</LangVersion>
+  </PropertyGroup>
   <ItemGroup>
```
Comment (Aramirez61) [inline Assignment/Assignment.csproj:6]: add this item group to preserve newest to follow the requirement Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌

<ItemGroup>
	  <None Update="People.csv">
	      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
	  </None>
</ItemGroup>

*indentation will need to be fixed, pasting into here changed it a bit

File: Assignment/SampleData.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            return People.Where(person => filter(person.EmailAddress))
+                .Select(person => (person.FirstName, person.LastName));
+        }
+
+        // 6.
+        public string GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)
+        {
+            var states = people.Select(person => person.Address.State)
+                .Distinct()
+                .OrderBy(state => state);
```
Comment (Aramirez61) [inline Assignment/SampleData.cs:80]: after ordering by state, make sure they are distinct, then after that, use the Aggregate function to instead of string.join to add the comma between states

File: Assignment/Node.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        Node<T> cur = this;
+        do
+        {
+            yield return cur;
+            cur = cur.Next;
+        } while (cur != this);
+    }
+
+    public IEnumerable<Node<T>> ChildValues(int max)
```
Comment (Aramirez61) [inline Assignment/Node.cs:83]: this should be ChildItems

**Review by Aramirez61 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
Using LINQ, skip the first row in the People.csv. ❌
-(Implement this in csv rows when it is being read in)

Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ❌✔
-(Not applicable)

Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ❌
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ❌
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ❌
Don't forget the list should be unique. ❌
-(Add a Distinct() method call)
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔


Overall, good job on implementing the SampleData and Node class. There are a couple requirements that I commented on that you can make changes to to more closely follow them. Make sure to unit test the SampleData class and the additions to the node class.

**Comment by MattA41** [comment]
im reviewing

File: Assignment/Assignment.csproj (hunk 4 above)
Comment (Aramirez61) [reply Assignment/Assignment.csproj:6]: for some reason the code i pasted isn't showing when I post the comment, but if you look at me and Cynthias Assignment.csproj on our PR you'll see the code i'm talking about.

**Review by MattA41 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ❌
        Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌ (Im not seeing this anywhere)
        Using LINQ, skip the first row in the People.csv. ❌(yall are not skipping the first line, just loading in the last line
        Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ❌✔

    Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
        Use ISampleData.CsvRows for your data source. ✔
        Don't forget the list should be unique. ✔
        Sort the list alphabetically. ✔
        Include a test that leverages a hardcoded list of addresses. ❌
        Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌

    Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
        Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
        Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
        Given the array, consider using string.Join to combine the list into a single string. ✔

    Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
        Use ISampleData.CsvRows as the source of the data. ✔
        Sort the list by State, City, and Zip. (Sort the addresses first then select). ❌ (you made the people before sorting by state)
        Be sure that Person.Address is also populated. ✔
        Adding null validation to all the Person and Address properties is optional.
        Consider using ISampleData.CsvRows in your test to verify your results. ❌

    Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
        Use ISampleData.People for your data source. ✔

    Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ❌(you are not joing the states into one string)
        Use the people parameter from ISampleData.People property for your data source. ✔
        At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ❌
        Don't forget the list should be unique. ✔
        It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

    Given the implementation of Node in Assignment5

    Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
    Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Extra Credit

    Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌✔

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.
    Place all shared project items into a Directory.Build.targets file.
    Ensure nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
    All of the above should be unit tested ❌
    Choose simplicity over complexity ✔

File: Assignment/SampleData.cs (hunk 2 above)
Comment (HadiM2) [reply Assignment/SampleData.cs:22]: You are correct

File: Assignment/SampleData.cs (hunk 3 above)
Comment (HadiM2) [reply Assignment/SampleData.cs:6]: just did, thank you!

File: Assignment/Assignment.csproj (hunk 4 above)
Comment (HadiM2) [reply Assignment/Assignment.csproj:6]: Thank you saw it!

File: Assignment/Node.cs (hunk 6 above)
Comment (HadiM2) [reply Assignment/Node.cs:83]: > this should be ChildItems

should act the same lol, thank you.

File: Directory.Build.props [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<Project>
+<PropertyGroup>
+	<TargetFramework>net8.0</TargetFramework>
+	<Nullable>enable</Nullable>
+	<LangVersion>12.0</LangVersion>
+	<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+	<EnableNETAnalyzers>true</EnableNETAnalyzers>
```
[the remark this answers is not in the document]
Comment (HadiM2) [reply Directory.Build.props:7]: how would I go about doing that?

File: Assignment.Tests/SampleDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+namespace Assignment.Tests;
+public class SampleDataTests
+{
+    [Fact]
+    public void SampleTest()
+    {
+        string str = "meow";
+        Assert.Equal("meow", str);
+    }
+}
```
[the remark this answers is not in the document]
Comment (HadiM2) [reply Assignment.Tests/SampleDataTests.cs]: Done now, thx for the reminder.

File: Assignment.Tests/Assignment.Tests.csproj [hunk 9]
```diff
@@ -0,0 +1,30 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+		<IsPackable>false</IsPackable>
+		<TargetFramework>net8.0</TargetFramework>
+		<LangVersion>12.0</LangVersion>
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:6]: Why are these here instead of just in the props file?

File: Assignment.Tests/NodeTests.cs [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        Assert.Equal("hello", node.Next.GetData());
+        node.Append("bye");
+        Assert.Equal("bye", node.Next.GetData());
+        Assert.Throws<ArgumentException>(() => node.Append("first"));
+        Assert.NotEqual("first", node.Next.GetData());
+        Assert.Equal("bye", node.Next.GetData());
+    }
+
+    [Fact]
+    public void toStringPrintToString()
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:96]:
```suggestion
    public void ToStringPrintToString()
```

File: Assignment.Tests/NodeTests.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using Xunit;
+
+namespace Assignment.Tests;
+
+public class NodeTests
+{
+    // NOTE: Had to remove Underscores from test method names as it was giving build error on git, we know they are needed but that us the solution to get passed that what I am guessing in unintended behavior.
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:8]: This is when you disable warning and justify it with that justification

File: Assignment/Node.cs (hunk 6 above)
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:83]: Why aren't their tests around this?

File: Assignment/SampleData.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Assignment;
+
+public class SampleData : ISampleData
+{
+    private const string FILEPATH = "People.csv";
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:10]: This is not how we name fields
